src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,168 @@
+            /// <param name="progressObject">The object implementing <see cref="IProgress{T}"/>.</param>
+            public ProgressParamInformation(object progressObject)
+            {
+                Requires.NotNull(progressObject, nameof(progressObject));
+
+                Type iProgressOfTType = MessageFormatterProgressTracker.FindIProgressOfT(progressObject.GetType());
+
+                Verify.Operation(iProgressOfTType != null, Resources.FindIProgressOfTError);
+
+                this.ValueType = iProgressOfTType.GenericTypeArguments[0];
+                this.ReportMethod = iProgressOfTType.GetRuntimeMethod(nameof(IProgress<int>.Report), new Type[] { this.ValueType });
+                this.ProgressObject = progressObject;
+            }
+
+            /// <summary>
+            /// Gets the actual <see cref="Type"/> reported by <see cref="IProgress{T}"/>.
+            /// </summary>
+            public Type ValueType { get; }
+
+            /// <summary>
+            /// Gets the <see cref="MethodInfo"/> of <see cref="IProgress{T}.Report(T)"/>.
+            /// </summary>
+            public MethodInfo ReportMethod { get; }
+
+            /// <summary>
+            /// Gets the instance of the object implementing <see cref="IProgress{T}"/>.
+            /// </summary>
+            public object ProgressObject { get; }
+        }
+
+        /// <summary>
+        /// Class that implements <see cref="IProgress{T}"/> and sends <see cref="ProgressRequestSpecialMethod"/> notification when reporting.
+        /// </summary>
+        public class JsonProgress<T> : IProgress<T>
+        {
+            private readonly JsonRpc rpc;
+            private readonly long? token;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="JsonProgress{T}"/> class.
+            /// </summary>
+            /// <param name="rpc">The <see cref="JsonRpc"/> instance used to send the <see cref="ProgressRequestSpecialMethod"/> norification.</param>
+            /// <param name="token">The <see cref="long"/> progress token used to obtain the <see cref="ProgressParamInformation"/> intance from <see cref="ProgressMap"/>.</param>
+            public JsonProgress(JsonRpc rpc, long? token)
+            {
+                this.rpc = rpc ?? throw new ArgumentNullException(nameof(rpc));
+                this.token = token ?? throw new ArgumentNullException(nameof(token));
+            }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="JsonProgress{T}"/> class.
+            /// </summary>
+            /// <param name="rpc">The <see cref="JsonRpc"/> instance used to send the <see cref="ProgressRequestSpecialMethod"/> norification.</param>
+            /// <param name="token">The <see cref="JToken"/> progress token used to obtain the <see cref="ProgressParamInformation"/> intance from <see cref="ProgressMap"/>.</param>
+            public JsonProgress(JsonRpc rpc, JToken token)
```

Comment:
>JToken [](start = 45, length = 6)

We should avoid `JToken` or any other newtonsoft.json type in shared infrastructure. Can you remove it?